src/ui/common/src/components/pages/artifact/id/index.tsx

```diff
@@ -1,204 +1,89 @@
-          workflowDagResultWithLoadingStatus?.result,
-          artifactId
-        )
-      : { metrics: [], checks: [] };
-
-  const pathPrefix = getPathPrefix();
-  const workflowLink = `${pathPrefix}/workflow/${workflowId}?workflowDagResultId=${workflowDagResultId}`;
-  const breadcrumbs = [
-    BreadcrumbLink.HOME,
-    BreadcrumbLink.WORKFLOWS,
-    new BreadcrumbLink(
-      workflowLink,
-      workflowDagResultWithLoadingStatus?.result?.name ?? 'Workflow'
-    ),
-    new BreadcrumbLink(path, artifact ? artifact.name : 'Artifact'),
-  ];
-
-  useEffect(() => {
-    // Load workflow dag result if it's not cached
-    if (
-      !workflowDagResultWithLoadingStatus ||
-      isInitial(workflowDagResultWithLoadingStatus.status)
-    ) {
-      dispatch(
-        handleGetWorkflowDagResult({
-          apiKey: user.apiKey,
-          workflowId,
-          workflowDagResultId,
-        })
-      );
-    }
-  }, [
-    dispatch,
-    user.apiKey,
+  const {
+    breadcrumbs: wfBreadcrumbs,
+    workflowId,
+    workflowDagId,
     workflowDagResultId,
+    workflowDagWithLoadingStatus,
     workflowDagResultWithLoadingStatus,
-    workflowId,
-  ]);
-
-  useEffect(() => {
-    if (!!artifact) {
-      if (!sideSheetMode) {
-        document.title = `${
-          artifact ? artifact.name : 'Artifact Details'
-        } | Aqueduct`;
-      }
+  } = useWorkflow(
+    user.apiKey,
+    workflowIdProp,
+    workflowDagIdProp,
+    workflowDagResultIdProp
+  );
 
-      if (
-        !!artifact.result &&
-        // intentional '==' to check undefined or null.
-        artifact.result.content_serialized == null &&
-        !contentWithLoadingStatus
-      ) {
-        dispatch(
-          handleGetArtifactResultContent({
-            apiKey: user.apiKey,
-            artifactId,
-            artifactResultId,
-            workflowDagResultId,
-          })
-        );
-      }
-    }
-  }, [
-    artifact,
-    artifactId,
-    artifactResultId,
-    contentWithLoadingStatus,
-    dispatch,
-    sideSheetMode,
+  const { breadcrumbs, artifact, contentWithLoadingStatus } = useArtifact(
     user.apiKey,
+    artifactIdProp,
+    wfBreadcrumbs,
     workflowDagResultId,
-  ]);
-
-  if (
-    !workflowDagResultWithLoadingStatus ||
-    isInitial(workflowDagResultWithLoadingStatus.status) ||
-    isLoading(workflowDagResultWithLoadingStatus.status)
-  ) {
-    return (
-      <Layout breadcrumbs={breadcrumbs} user={user}>
-        <CircularProgress />
-      </Layout>
-    );
-  }
+    workflowDagWithLoadingStatus,
+    workflowDagResultWithLoadingStatus,
+    !sideSheetMode
+  );
 
-  if (isFailed(workflowDagResultWithLoadingStatus.status)) {
-    return (
-      <Layout breadcrumbs={breadcrumbs} user={user}>
-        <Alert severity="error">
-          <AlertTitle>Failed to load workflow.</AlertTitle>
-          {workflowDagResultWithLoadingStatus.status.err}
-        </Alert>
-      </Layout>
-    );
-  }
+  const dagResult =
```

Comment:
this ensures we fall back to `wfDagWithLoadingStatus` to retrieve various objects (e.g. up / down stream) when `dagResults` is not available.